### PR TITLE
Reduce memory footprint of `PhpStormStubsSourceStubber` and `MemoizingParser`

### LINKED ISSUE #624: Tests consuming alot memory right now
At the moment the test are consuming a huge amount of memory. Local the test are consuming 2.5GB:

```
$ vendor/bin/phpunit
PHPUnit 9.1.5 by Sebastian Bergmann and contributors.

.............................................................   61 / 7358 (  0%)
.............................................................  122 / 7358 (  1%)
.............................................................  183 / 7358 (  2%)
.............................................................  244 / 7358 (  3%)
.............................................................  305 / 7358 (  4%)
.............................................................  366 / 7358 (  4%)
.............................................................  427 / 7358 (  5%)
.............................................................  488 / 7358 (  6%)
.............................................................  549 / 7358 (  7%)
.............................................................  610 / 7358 (  8%)
.............................................................  671 / 7358 (  9%)
.............................................................  732 / 7358 (  9%)
.............................................................  793 / 7358 ( 10%)
.............................................................  854 / 7358 ( 11%)
.............................................................  915 / 7358 ( 12%)
.............................................................  976 / 7358 ( 13%)
............................................................. 1037 / 7358 ( 14%)
............................................................. 1098 / 7358 ( 14%)
...............................................S............. 1159 / 7358 ( 15%)
............................................................. 1220 / 7358 ( 16%)
............................................................. 1281 / 7358 ( 17%)
............................................................. 1342 / 7358 ( 18%)
............................................................. 1403 / 7358 ( 19%)
............................................................. 1464 / 7358 ( 19%)
............................................................. 1525 / 7358 ( 20%)
............................................................. 1586 / 7358 ( 21%)
............................................................. 1647 / 7358 ( 22%)
............................................................. 1708 / 7358 ( 23%)
............................................................. 1769 / 7358 ( 24%)
............................................................. 1830 / 7358 ( 24%)
............................................................. 1891 / 7358 ( 25%)
............................................................. 1952 / 7358 ( 26%)
............................................................. 2013 / 7358 ( 27%)
............................................................. 2074 / 7358 ( 28%)
............................................................. 2135 / 7358 ( 29%)
............................................................. 2196 / 7358 ( 29%)
............................................................. 2257 / 7358 ( 30%)
............................................................. 2318 / 7358 ( 31%)
............................................................. 2379 / 7358 ( 32%)
............................................................. 2440 / 7358 ( 33%)
............................................................. 2501 / 7358 ( 33%)
............................................................. 2562 / 7358 ( 34%)
............................................................. 2623 / 7358 ( 35%)
............................................................. 2684 / 7358 ( 36%)
............................................................. 2745 / 7358 ( 37%)
............................................................. 2806 / 7358 ( 38%)
............................................................. 2867 / 7358 ( 38%)
............................................................. 2928 / 7358 ( 39%)
............................................................. 2989 / 7358 ( 40%)
............................................................. 3050 / 7358 ( 41%)
............................................................. 3111 / 7358 ( 42%)
............................................................. 3172 / 7358 ( 43%)
............................................................. 3233 / 7358 ( 43%)
............................................................. 3294 / 7358 ( 44%)
............................................................. 3355 / 7358 ( 45%)
............................................................. 3416 / 7358 ( 46%)
............................................................. 3477 / 7358 ( 47%)
............................................................. 3538 / 7358 ( 48%)
............................................................. 3599 / 7358 ( 48%)
............................................................. 3660 / 7358 ( 49%)
............................................................. 3721 / 7358 ( 50%)
............................................................. 3782 / 7358 ( 51%)
............................................................. 3843 / 7358 ( 52%)
............................................................. 3904 / 7358 ( 53%)
............................................................. 3965 / 7358 ( 53%)
............................................................. 4026 / 7358 ( 54%)
............................................................. 4087 / 7358 ( 55%)
............................................................. 4148 / 7358 ( 56%)
............................................................. 4209 / 7358 ( 57%)
............................................................. 4270 / 7358 ( 58%)
............................................................. 4331 / 7358 ( 58%)
............................................................. 4392 / 7358 ( 59%)
............................................................. 4453 / 7358 ( 60%)
............................................................. 4514 / 7358 ( 61%)
............................................................. 4575 / 7358 ( 62%)
............................................................. 4636 / 7358 ( 63%)
............................................................. 4697 / 7358 ( 63%)
............................................................. 4758 / 7358 ( 64%)
............................................................. 4819 / 7358 ( 65%)
............................................................. 4880 / 7358 ( 66%)
............................................................. 4941 / 7358 ( 67%)
............................................................. 5002 / 7358 ( 67%)
............................................................. 5063 / 7358 ( 68%)
............................................................. 5124 / 7358 ( 69%)
............................................................. 5185 / 7358 ( 70%)
............................................................. 5246 / 7358 ( 71%)
............................................................. 5307 / 7358 ( 72%)
............................................................. 5368 / 7358 ( 72%)
............................................................. 5429 / 7358 ( 73%)
............................................................. 5490 / 7358 ( 74%)
............................................................. 5551 / 7358 ( 75%)
............................................................. 5612 / 7358 ( 76%)
............................................................. 5673 / 7358 ( 77%)
............................................................. 5734 / 7358 ( 77%)
............................................................. 5795 / 7358 ( 78%)
............................................................. 5856 / 7358 ( 79%)
............................................................. 5917 / 7358 ( 80%)
............................................................. 5978 / 7358 ( 81%)
............................................................. 6039 / 7358 ( 82%)
............................................................. 6100 / 7358 ( 82%)
............................................................. 6161 / 7358 ( 83%)
............................................................. 6222 / 7358 ( 84%)
............................................................. 6283 / 7358 ( 85%)
............................................................. 6344 / 7358 ( 86%)
............................................................. 6405 / 7358 ( 87%)
............................................................. 6466 / 7358 ( 87%)
............................................................. 6527 / 7358 ( 88%)
............................................................. 6588 / 7358 ( 89%)
............................................................. 6649 / 7358 ( 90%)
............................................................. 6710 / 7358 ( 91%)
............................................................. 6771 / 7358 ( 92%)
............................................................. 6832 / 7358 ( 92%)
............................................................. 6893 / 7358 ( 93%)
............................................................. 6954 / 7358 ( 94%)
............................................................. 7015 / 7358 ( 95%)
............................................................. 7076 / 7358 ( 96%)
............................................................. 7137 / 7358 ( 96%)
............................................................. 7198 / 7358 ( 97%)
............................................................. 7259 / 7358 ( 98%)
............................................................. 7320 / 7358 ( 99%)
......................................                        7358 / 7358 (100%)

Time: 00:24.055, Memory: 2.50 GB
```

I just checked some CI jobs from the last days:
2.55 GB: https://github.com/Roave/BetterReflection/runs/717137271?check_suite_focus=true
1.41 GB: https://github.com/Roave/BetterReflection/pull/622/checks?check_run_id=716753844
212 MB: https://github.com/Roave/BetterReflection/pull/607/checks?check_run_id=714702198

Right now i could not pin down the culprit but I will investigate further..

### PR DESCRIPTION
**Pull request by kukulich**
Fixes #624

### CODE REVIEW

**Comment by ondrejmirtes** [comment]
Still more than double than before I guess :( @Ocramius is right that this should be supported but if I had to choose between:

1) Be correct for a small percentage of users that would take advantage of case-insensitive constants and make the library worse for everyone
2) Be a little bit incorrect (that we were this way until 4.3) and consume much less memory

...it's obvious what I'd choose :)

**Comment by kukulich** [comment]
The rest of the memory usage is probably not related to case insensitive constants.

PR builds: https://github.com/Roave/BetterReflection/actions?query=branch%3Aphpstorm-stubber
The last one is without static properties and the memory usage changed from 216.00 MB to 1.41 GB.

Current PR uses static properties again and it decreased from 2.31 GB to 1.38 GB: https://github.com/Roave/BetterReflection/pull/627/checks?check_run_id=718176772

I've tried revert of case insensitive constant just to test it and it's even worse - 1.85 GB: https://github.com/Roave/BetterReflection/pull/628/checks?check_run_id=718188887

**Comment by kukulich** [comment]
1.41 GB - Already contains case insensitive constants.
https://github.com/Roave/BetterReflection/runs/717048810?check_suite_focus=true

After #606 has been merged: 2.55 GB
https://github.com/Roave/BetterReflection/runs/717069119?check_suite_focus=true

So it looks it's something in #606.

**Comment by Ocramius** [comment]
Can you try reverting my changes to the memoizing source locator? Maybe
that had a bigger impact than expected.

**Comment by kukulich** [comment]
@Ocramius Tested just now. It decresed to 178.00 MB (this PR + the revert). Probably problems with serialize/unserialize.

**Comment by kukulich** [comment]
It also faster: 12 second vs 18 seconds on my ntb

**Comment by Ocramius** [comment]
Yeh, strings are heavy. We can run with the revert, but I'll need to also ensure that a few of the tests run in a separate process.

Also unsure about reverting the constraint that guarantees distinct AST nodes...

**Comment by ondrejmirtes** [comment]
BTW what we did in CachedParser in PHPStan (same concept as Memoizing) some time ago is that we don't keep all cached keys in memory at once. There's an upper limit and keys get evicted when newer ones appear. Surprisingly this has minimal performance impact, but memory footprint is smaller. See: https://github.com/phpstan/phpstan-src/blob/master/src/Parser/CachedParser.php

**Comment by ondrejmirtes** [comment]
Funny realization for me - there's zero memory impact if I use `$sourceCode` or `hash('sha256', $sourceCode)` as array keys as there's probably some hashing function on the PHP side applied anyway.

**Comment by kukulich** [comment]
I've improved the memory usage with `json_encode` instead of `serialize`. It's not perfect, it's just better.

**Comment by kukulich** [comment]
The last commit decreased the memory usage to 294.00 MB but it's a little ugly...

**Comment by kukulich** [comment]
The solution here is not optimal (it's a little ugly) but has better memory usage than `serialize/unserialize`

File: src/SourceLocator/Ast/Parser/MemoizingParser.php [hunk 1]
```diff
@@ -4,27 +4,42 @@
...
+
+    private Closure $decode;
+
     public function __construct(Parser $wrappedParser)
     {
         $this->wrappedParser = $wrappedParser;
+        $this->jsonDecoder   = new JsonDecoder();
+        $this->decode        = Closure::bind(function (array $value) {
+            /** @psalm-suppress UndefinedMethod */
+            return $this->decodeRecursive($value);
```
Comment (Ocramius) [inline src/SourceLocator/Ast/Parser/MemoizingParser.php:41]: Hmm, no, we really shouldn't do this.

This is `private` API of `PhpParser` that we shouldn't touch at all: doing a `json_encode()` before passing a structure to the decoder is preferable, even if it comes with a performance impact.

Alternatively, we port `JsonDecoder` (with tests) completely, but overall we could do something much simpler:

 1. deprecate `MemoizingParser` and leave it as-is
 2. create a new version of `MemoizingParser` that keeps AST nodes in-memory as they are (no cloning nor anything), but we do mention that explicitly in naming and documentation somehow (that it isn't safe to change the AST)

Comment (kukulich) [reply src/SourceLocator/Ast/Parser/MemoizingParser.php:41]: Understood :) I can remove the last commit. The first two improved the memory too.

Comment (kukulich) [reply src/SourceLocator/Ast/Parser/MemoizingParser.php]: Done.

**Comment by Ocramius** [comment]
Thanks @kukulich!

**Pull request merged**